apps/metric_processors.go

```diff
@@ -36,13 +36,13 @@ func (p MetricsProcessorExcludeMetrics) Type() string {
 func (p MetricsProcessorExcludeMetrics) Processors() []otel.Component {
 	var metricNames []string
 	for _, glob := range p.MetricsPattern {
-		// TODO: Remove TrimPrefix when we support metrics with other prefixes.
-		glob = strings.TrimPrefix(glob, "agent.googleapis.com/")
+		// $ needs to be escaped because reasons.
+		// https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/metricstransformprocessor#rename-multiple-metrics-using-substitution
 		var literals []string
 		for _, g := range strings.Split(glob, "*") {
-			literals = append(literals, regexp.QuoteMeta(g))
+			literals = append(literals, strings.Replace(regexp.QuoteMeta(g), "$", "$$", -1))
```

Comment:
I think filterprocessor doesn't need to escape $

Reply:
Why did you remove this escaping? `$` is special in the otel configuration; it starts an environment variable expansion.

Reply:
when I tested, ^$ works just fine. What’s the context that it needs to be escaped as $$? I saw metricstransformer uses $$ in recent changes. I don’t be think filterprocessor needs it.

Reply:
I just confirmed again it works just fine with a single $.

```
curl -sSO https://dl.google.com/cloudagents/add-google-cloud-ops-agent-repo.sh
sudo bash add-google-cloud-ops-agent-repo.sh --also-install
sudo service google-cloud-ops-agent.service stop
sudo vim /run/google-cloud-ops-agent-opentelemetry-collector/otel.yaml <---modify this file to exclude `^processes/.*$` and let it output to logging exporter
sudo /opt/google-cloud-ops-agent/subagents/opentelemetry-collector/otelopscol --add-instance-id=false --config=/run/google-cloud-ops-agent-opentelemetry-collector/otel.yaml
```

Reply:
I've also successfully tested with the single `$` — it looks like only metricstransform needed the escaping.

Reply:
@sophieyfang I don't think your test is valid because `^agent.*processes/.*$` and `^agent.*processes/.*` have the same behavior. 

@igorpeshansky What exactly did you test? I would expect `^agent.*processes/c` to exclude `agent/processes/count_by_state` but `^agent.*processes/c$` to not. So if it's treating the latter as the former, that would be the behavior change.

Also it's possible a trailing $ is treated specially; the other case to test is `^workload.*foo\$bar`, which needs to match a metric named `foo$bar` (and not just be a failure to match anything). The latter is what the `strings.Replace` is for.

Reply:
Tested again with:
“””
filter/default__pipeline_hostmetrics_3:
    metrics:
      include:
        match_type: regexp
        metric_names:
        - ^agent.googleapis.com/processes/disk/read_bytes_count$
 “””
It does only outputread_bytes_count metrics when I tested it. What error/mis-behavior is expected by this expanding of $? I don’t see it emit any error or stop outputting metrics

I agree that .*$ works the same as .*

Reply:
I've tested the code that generates a single `$` within the current constraints, which means that the regexp ended in a `.*`. The first example you gave isn't valid as of today. The second example (`foo$bar`) is harder to test, since no such metric is generated.

I've just tried something like this: `metrics_pattern: ["agent.googleapis.com/processes$XYZ/*"]`, and there's definitely substitution going on, as the following got logged:
```
Nov 24 19:27:06 b-205077265 otelopscol[16397]: 2021-11-24T19:27:06.139Z        info        filterprocessor@v0.37.1-0.20211028205244-e6fab4102b84/filter_processor.go:73        Metric filter configured        {"kind": "processor", "name": "filter/default__pipeline_hostmetrics_3", "include match_type": "", "include expressions": [], "include metric names": [], "include metrics with resource attributes": null, "exclude match_type": "regexp", "exclude expressions": [], "exclude metric names": ["^agent\\.googleapis\\.com/processes\\/.*$"], "exclude metrics with resource attributes": null}
```

The trailing `$` is indeed treated differently by `os.Expand`, which explains why our tests worked. I've reverted to the commit that does the escaping.

Reply:
$$ sounds good to me